<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:macro name="diffbot_ros2_control" params="name prefix">

    <ros2_control name="${name}" type="system">
      <hardware>
        <plugin>dfrobot_dc_motor_hardware/DFRobotDCMotorHardware</plugin>
        <param name="left_wheel_name">left_wheel_joint</param>
        <param name="right_wheel_name">right_wheel_joint</param>
        <param name="loop_rate">30</param>
        <param name="bus_id">1</param> <!-- I2C Bus number (e.g., 1 for /dev/i2c-1) -->
        <param name="i2c_address">16</param> <!-- I2C Address of DFRobot board -->
        <param name="max_rpm">75</param>
        <param name="encoder_reduction_ratio">210</param>
        <param name="timeout_ms">1000</param>
        <!-- <param name="enc_counts_per_rev">3436</param>
        <param name="pid_p">20</param>
        <param name="pid_d">12</param>
        <param name="pid_i">0</param>
        <param name="pid_o">50</param> -->
      </hardware>
      <joint name="${prefix}left_wheel_joint">
        <command_interface name="velocity"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="${prefix}right_wheel_joint">
        <command_interface name="velocity"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
    </ros2_control>

  </xacro:macro>

</robot>
